src/lib/info_tree.rs

```diff
@@ -0,0 +1,544 @@
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+#![allow(dead_code)] // Temporary, this module will be used in RTED.
+
+use ast::{Arena, NodeId};
+use label_maps::LabelMap;
+/// Stores information about a single AST (as needed by RTED) in vectors.
+use std::ops::{Index, IndexMut};
+
+// Size of the `info` vector.
+const INFO_SIZE: usize = 16;
+// Size of the paths vector.
+const PATHS_SIZE: usize = 3;
+// Size of the relative subtrees vector.
+const REL_SUBTREES_SIZE: usize = 3;
+// Size of the node type vector.
+const NODE_TYPE_SIZE: usize = 3;
+
+enum PathIdx {
+    Left = 0,
+    Right = 1,
+    Heavy = 2,
+    Both = 3
+}
+
+/// Index a struct with the `PathIdx` enum.
+macro_rules! table_index_trait {
+    ($table:ident, $ty:ty) => {
+        impl Index<PathIdx> for $table {
+            type Output = $ty;
+            fn index(&self, idx: PathIdx) -> &Self::Output {
+                &self.table[idx as usize]
+            }
+        }
+        impl IndexMut<PathIdx> for $table {
+            fn index_mut(&mut self, idx: PathIdx) -> &mut Self::Output {
+                &mut self.table[idx as usize]
+            }
+        }
+    };
+}
+
+// Flags (left, right, heavy) for each node.
+struct NodeTypeTable {
+    pub table: Vec<Vec<bool>>
+}
+
+impl NodeTypeTable {
+    fn new(size: usize) -> NodeTypeTable {
+        let mut node_types = NodeTypeTable { table: vec![] };
```

Comment:
These `debug_assert_eq`s also look pointless to me, unless I'm missing out on some subtlety.

Reply:
It's not subtle, I was just concerned about off by one errors and it seemed a good way to check. No need to merge it into master though.

Reply:
All the ones I highlighted look to me to be correct by construction within the function. I think they probably clutter things up more than they help.